NixOS and xmonad set wallpaper script compatibility fixes

willfish opened the pull request:
What

- Added a desktop detection function/tests
- Use this function in the set_wallpaper script
- Wrap 100% of DE in a if/elif branch with a fallback of simple wm handling with feh/nitrogen
- Adds some convenience nix utilities for myself
- Update to using `!#/usr/bin/env bash` for support for nixos

Why

- nixos adds a none prefix to the variables the set wallpaper script is checking (none+xmonad) which breaks variety for simple wms
- /bin/bash isn't available on nixos by default

AC

- Tested running the script using real wallpapers downloaded by variety
- Ran the bats tests

willfish commented on data/scripts/set_wallpaper:
```diff
@@ -227,17 +227,35 @@ fi
+)
+
+for wm in "${SIMPLE_WMS[@]}"; do
+  if [[ "$XDG_CURRENT_DESKTOP $XDG_SESSION_DESKTOP $DESKTOP_SESSION" = *"$wm"* ]]; then
```
So this concatenates all of these variables and does a substring check against our simple wm variable.

For example:

`"none+xmonad none+xmonad none+xmonad"` would substring match against `xmonad`

willfish commented:
@jlu5 Do you mind if we cut a new release after this change? My package manager essentially depends on downloading the 0.8.13.tar.gz file from the following URL: https://github.com/varietywalls/variety/archive/refs/tags/0.8.13.tar.gz

willfish commented:
If anyone that uses nix come into this issue with not setting the wallpaper they can add the following activation script to get this working:

```nix
  system.activationScripts.binBash = ''
    mkdir -p /bin
    ln -sf ${pkgs.bash}/bin/bash /bin/bash
  '';
```

jlu5 commented on variety_lib/varietyconfig.py:
```diff
@@ -23,7 +23,7 @@
     __variety_data_directory__ = "../data"
 __license__ = "GPL-3"
-__version__ = "0.8.12"
+__version__ = "0.8.13"
```
The app version should not be changed in a PR. This needs to be done separately (with corresponding changelog updates) and tagged by a maintainer

willfish replied:
Okey doke. Thanks for the feedback

jlu5 commented on data/scripts/set_wallpaper:
```diff
@@ -227,17 +227,48 @@ fi
+# The syntax for these is <desktop>+<window_manager>.
+#
+# Simple window managers usually have no desktop environment
+# so the init script should set XDG_CURRENT_DESKTOP to "none+i3" or similar.
```
Sorry, is this actually an established convention? I don't see the "none+" mentioned in https://wiki.archlinux.org/title/Xdg-utils#Environment_variables for instance, and I find it awkwardly redundant.

willfish replied:
My understanding is it's part of the display manager and specific to nixos. You're idea to try setting with feh or nitrogen would mean we don't have to know any of this so probably that is the route to take

jlu5 reviewed (commented):
Honestly, instead of hardcoding a list of WMs, I wonder if set_wallpaper should unconditionally try feh and nitrogen if it fails to detect any known system. This would probably mean that the script needs to early return when it does detect a DE, but this would altogether simplify the code here.

willfish commented:
@jlu5 I've simplified and incorporated your feedback and have tested this for myself on NixOS with xmonad as my wm and its all working. This is actually more idiomatic to the way a bunch of other commands get executed in the file so that was a great suggestion. I've also reverted the version adjustment

jlu5 commented:
I think we should guard the feh and nitrogen blocks to only run if another DE wasn't detected. It doesn't seem right to have two tools try to set the wallpaper at the same time, for example.

willfish commented:
Re DE detection:

We specifically moved away from that

> Honestly, instead of hardcoding a list of WMs, I wonder if set_wallpaper should unconditionally try feh and nitrogen if it fails to detect any known system. This would probably mean that the script needs to early return when it does detect a DE, but this would altogether simplify the code here.

Re using both feh and nitrogen (assuming they're both installed on the system):

This has no impact since we exit 0 when one was detected and executed

```bash
# For simple WMs, use either feh or nitrogen
# TODO: These should take the scaling parameter $4 into account and use other options than --bg-fill
if command -v "feh" >/dev/null 2>&1; then
    feh --bg-fill "$WP" 2> /dev/null
    exit 0
elif command -v "nitrogen" >/dev/null 2>&1; then
    nitrogen --set-zoom-fill --save "$WP" 2> /dev/null
    exit 0
fi
```

So I'm setting two values in my nix config right now to make variety work for simple window managers on nixos:

1. Making sure the `DESKTOP_SESSION` value is set in a compatible way with your script https://github.com/willfish/nix/blob/master/home/config/xmonad/xmonad.hs#L79
2. Making sure that `/bin/bash` is available on the path https://github.com/willfish/nix/blob/master/system/modules/common-configuration.nix#L3-L6

My change only fixes 1 and I'm wondering whether you can't think of a standard way to fix 2 without needing nix users to change their activation scripts?

Could we move to `/bin/sh` for example or extend the README.md to include the shell activation example I"ve used? Less realistically we could make the scripts fully POSIX compliant?

jlu5 commented:
> Re DE detection:
> 
> We specifically moved away from that
> 
> > Honestly, instead of hardcoding a list of WMs, I wonder if set_wallpaper should unconditionally try feh and nitrogen if it fails to detect any known system. This would probably mean that the script needs to early return when it does detect a DE, but this would altogether simplify the code here.
> 
> Re using both feh and nitrogen (assuming they're both installed on the system):
> 
> This has no impact since we exit 0 when one was detected and executed

I'm more concerned how feh or nitrogen would interact with a DE that we already support setting the wallpaper for (GNOME, KDE, Xfce, etc.). It's possible they're installed in addition to a well-known DE, so the blocks that set the wallpaper for those should also early exit. Right now, there are lot of cases that blindly set a value without checking the current DE, so those should ideally be fixed as well...

>     1. Making sure the `DESKTOP_SESSION` value is set in a compatible way with your script https://github.com/willfish/nix/blob/master/home/config/xmonad/xmonad.hs#L79
> 
>     2. Making sure that `/bin/bash` is available on the path https://github.com/willfish/nix/blob/master/system/modules/common-configuration.nix#L3-L6
> 
> 
> My change only fixes 1 and I'm wondering whether you can't think of a standard way to fix 2 without needing nix users to change their activation scripts?

Does `/usr/bin/env bash` work?

> 
> Could we move to `/bin/sh` for example or extend the README.md to include the shell activation example I"ve used? Less realistically we could make the scripts fully POSIX compliant?

Bash scripting is less error-prone, and I'm willing to make a portability tradeoff here since we mainly target Linux systems anyways. See https://github.com/varietywalls/variety/issues/23

willfish commented:
> Does /usr/bin/env bash work?

It does - I just updated the script to reflect that fact. When I think about it the script will basically be compatible with any version of bash and `/usr/bin/env bash` makes loads more sense.

> I'm more concerned how feh or nitrogen would interact with a DE that we already support setting the wallpaper for (GNOME, KDE, Xfce, etc.). It's possible they're installed in addition to a well-known DE, so the blocks that set the wallpaper for those should also early exit. 

That makes sense. I've moved feh/nitrogen to the bottom of the script to take advantage of the DEs that are exiting properly.

Putting it all together:

| Desktop Environment                | Method / Tool                                                                                                                       | Exit Code Behavior                                                                                                              |
|------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------|
| **Enlightenment / Moksha**         | Edje template + `enlightenment_remote`                                                                                               | Continues through script; final exit is `0`                                                                                     |
| **KDE (Plasma 5.7+ / KDE 6)**        | DBus script via `dbus-send` (plus `kwriteconfig6` for KDE 6 lockscreen)                                                                | Exits immediately with the DBus command’s exit code (i.e. `0` if successful, non‑zero if not)                                   |
| **KDE (Older versions)**           | Copies the wallpaper file to a dedicated slideshow folder                                                                            | Falls through; final exit is `0`                                                                                                 |
| **GNOME 3 / Unity / Deepin**       | Uses `gsettings` to update desktop & screensaver (lock screen) wallpaper settings                                                     | Falls through; final exit is `0`                                                                                                 |
| **XFCE**                         | Uses `xfconf-query` to update wallpaper properties for each screen/monitor                                                             | Falls through; final exit is `0`                                                                                                 |
| **Lingmo OS**                      | Calls `qdbus` (via `com.lingmo.Settings`) to set the wallpaper                                                                          | Falls through; final exit is `0`                                                                                                 |
| **LXDE**                         | Uses `pcmanfm --set-wallpaper`                                                                                                         | Falls through; final exit is `0`                                                                                                 |
| **LXQt**                         | Uses `pcmanfm-qt --set-wallpaper`                                                                                                      | Falls through; final exit is `0`                                                                                                 |
| **Sway**                           | Uses `swaybg` (with a fallback to `swaymsg`) to set the wallpaper                                                                      | Falls through; final exit is `0`                                                                                                 |
| **Trinity**                        | Uses `dcop` to invoke `KBackgroundIface setWallpaper`                                                                                 | Falls through; final exit is `0`                                                                                                 |
| **MATE**                           | Uses `gsettings` to set `org.mate.background` wallpaper settings                                                                       | Falls through; final exit is `0`                                                                                                 |
| **Cinnamon**                       | Uses `gsettings` to set `org.cinnamon.desktop.background` wallpaper settings                                                           | Falls through; final exit is `0`                                                                                                 |
| **Awesome WM**                     | Sends a Lua command via `awesome-client` to maximize the wallpaper on all screens                                                      | Falls through; final exit is `0`                                                                                                 |
| **Simple Window Managers**         | Uses `feh` (or, if unavailable, `nitrogen`) to set the wallpaper                                                                       | Uses an explicit `exit 0` after calling the command                                                                              |

I guess we can probably just add an exit 0 when we've set these successfully.

willfish commented:
Okay, I've updated the script to exit if the last evaluated wallpaper command for each DE was successful.

This isn't true for GNOME since the usual variables for detecting whether we're on GNOME don't seem to be available in the script. Not sure if you think this looks okay. I can just revert and we can just leave feh/nitrogen doubly setting the wallpaper if that's preferable

willfish commented:
Curious what your sense is @jlu5

willfish commented:
If you prefer @jlu5 I can revert the change to do early returns and handle the other DEs and we can do that separately?

jlu5 commented:
IMO we ought to switch set_wallpaper's DE detection to be more explicit, rather than checking the return code of all the gsettings commands etc. It's possible for a dconf option to exist just because a DE's files are installed, even if it is not the current running desktop.

[get_wallpaper](https://github.com/varietywalls/variety/blob/master/data/scripts/get_wallpaper) does this in a much more straightforward way

willfish commented:
@jlu5 Let me know what you think of my latest commit.

I've:

1. Added a detect_desktop script which I've added a bats test for (ideally this would be part of a github actions workflow/equivalent which would be free)
2. Exhausted as far as I can make out the kinds of desktop environment we're interested in
3. Added if/elif branches and a fallback on feh/nitrogen to enable working more simply with a standard simple desktop environment

willfish commented:
![image](https://github.com/user-attachments/assets/b53a0977-5844-475a-af1f-bd2c95525c29)

willfish commented:
Could you run the set_wallpaper script on your local machine @jlu5?

willfish commented:
@jlu5 I could have just deleted the bits you didn’t like.

jlu5 commented:
Don't stress about it! I just pulled the parts and rearranged some things to avoid the extra back and forth

willfish commented:
Fair enough. I know it was a bit bold introducing bats but I didn’t trust myself to get that part right without some tests. I was actually debating just organising a zoom call as we’ve had a lot of back and forth. Appreciated your feedback

jlu5 commented:
I wouldn't put too much faith in asynchronous comms for an open source project. Everyone works on different timezones with (purposely) unpredictable schedules. (This is one of many hobby projects for me after all...)

To give you more context on the other parts I've changed:
- I removed the parts checking DESKTOP_SESSION. These days, XDG_CURRENT_DESKTOP is part of the standard spec and widely implemented in full desktop environments. DESKTOP_SESSION can instead reflect a display manager's session name, so you end up with awkward cases like "Lubuntu" which don't neatly map to one DE
- I inlined the DE detection for now to minimize the changes required. Otherwise to make this work, we'd have to update https://github.com/varietywalls/variety/blob/0.8.13/variety/VarietyWindow.py#L282 as well, as that code is what actually copies the scripts into `~/.config/variety`

PerchunPak commented on data/scripts/set_wallpaper:
```diff
@@ -139,162 +112,144 @@ if [ "${KDE_FULL_SESSION}" == "true" ]; then
-        kill "$PID" 2>/dev/null
+    echo '$''full ''$'"full|$WP||:0.0" >~/.fluxbox/lastwallpaper
+
+elif [ "$DE" == "sway" ]; then
```
On window managers, user can replace almost all components with whatever they like. This is why swaybg can be used independently of using sway. I use swaybg on hyprland, and this broke my setup ([hyprland is not supported by variety](https://wiki.hyprland.org/Useful-Utilities/Wallpapers/#hyprpaper), but [swaybg is](https://github.com/varietywalls/variety/pull/618/files), which works perfectly fine on hyprland)

Instead, you should check if swaybg command is available

CC @willfish @jlu5

PerchunPak replied:
Even if someone would add support for hyprpaper, I like swaybg more, and want to continue using it
